mlir-sycl/include/mlir/Dialect/SYCL/Analysis/ConstructorBaseAnalysis.h

```diff
@@ -0,0 +1,158 @@
+/// a merge point.
+///
+template <typename ConcreteAnalysis, typename AnalysisResult>
+class ConstructorBaseAnalysis {
+public:
+  ConcreteAnalysis &initialize(bool useRelaxedAliasing = false) {
+
+    // Initialize the dataflow solver
+    aliasAnalysis = &am.getAnalysis<mlir::AliasAnalysis>();
+    aliasAnalysis->addAnalysisImplementation(
+        sycl::AliasAnalysis(useRelaxedAliasing));
+
+    solver = std::make_unique<polygeist::DataFlowSolverWrapper>(*aliasAnalysis);
+
+    // Populate the solver and run the analyses needed by this analysis.
+    solver->loadWithRequiredAnalysis<polygeist::ReachingDefinitionAnalysis>(
+        *aliasAnalysis);
+
+    if (failed(solver->initializeAndRun(operation))) {
+      operation->emitError("Failed to run required dataflow analyses");
+      return *static_cast<ConcreteAnalysis *>(this);
+    }
+
+    // Let the derived class perform more initialization of necessary.
+    static_cast<ConcreteAnalysis *>(this)->finalizeInitialization(
+        useRelaxedAliasing);
+
+    initialized = true;
+
+    return *static_cast<ConcreteAnalysis *>(this);
+  }
+
+  ConstructorBaseAnalysis(Operation *op, AnalysisManager &mgr)
+      : operation{op}, am{mgr}, solver{nullptr}, aliasAnalysis{nullptr} {}
+
+protected:
+  template <typename SYCLType>
+  std::optional<AnalysisResult> getInformationFromConstruction(Operation *op,
+                                                               Value operand) {
+    assert(initialized &&
+           "Analysis only available after successful initialization");
+    assert(isa<LLVM::LLVMPointerType>(operand.getType()) &&
+           "Expecting an LLVM pointer");
+    assert(aliasAnalysis != nullptr && "Alias analysis not initialized");
+
+    const polygeist::ReachingDefinition *reachingDef =
+        solver->lookupState<polygeist::ReachingDefinition>(op);
+    assert(reachingDef && "expected a reaching definition");
+
+    auto mods = reachingDef->getModifiers(operand, *solver);
+    if (!mods || mods->empty())
+      return std::nullopt;
+
+    if (!llvm::all_of(*mods, [&](const polygeist::Definition &def) {
+          return isConstructor<SYCLType>(def);
+        }))
+      return std::nullopt;
+
+    auto pMods = reachingDef->getPotentialModifiers(operand, *solver);
+    if (pMods) {
+      if (!llvm::all_of(*pMods, [&](const polygeist::Definition &def) {
+            return isConstructor<SYCLType>(def);
+          }))
+        return std::nullopt;
+    }
+
+    bool first = true;
+    AnalysisResult info;
+    for (const polygeist::Definition &def : *mods) {
+      if (first) {
+        info = getInformation<SYCLType>(def);
+        first = false;
+      } else {
+        info = info.join(getInformation<SYCLType>(def), *aliasAnalysis);
+        if (info.isTop())
+          // Early return: As soon as joining the different information has
+          // reached the top of the lattice, we can end processing.
+          return info;
+      }
+    }
```

Comment:
This has a small flaw: if the first information obtained is top, we will still calculate the second one. Can we change fix this small thing?